### DISCUSSION TITLE
Visual testing

### DISCUSSION THREAD

**Discussion by Nerwen18** [selenide/selenide#2887] [General]
Hello everyone,

What visual testing frameworks are you using to test your code with selenide?
Looking for a tool/library from your experience that is suitable, thanks in advance!

**Comment by asolntsev** [comment]
Hi @Nerwen18 .
I personally don't use visual testing. I believe in automated testing. ;)

**Reply by Nerwen18** [reply]
Hello @asolntsev,
Thanks for your reply! :)
And how you handle situations like checking if a button is in the right place or any other visual characteristics, like color, shape, size?

**Reply by asolntsev** [reply]
Manually.
Somebody always clicks the application anyway. At least product owner when he tries the new functionality. 

Things like color and shape are hardly automatable. It's not reasonable to waste time on such automation that causes endless problems, flakiness and maintenance nightmare.

**Comment by thedefside** [comment]
I use visual testing with Selenide. I have found several styling related issues that a human likely would not have caught right away. Humans just don't have precision and patience to look at EVERYTHING.

Here's how I do it:

1. I use selenide-full-screenshot plugin and capture screenshot of each page in my application
2. I use mock services to make sure the data never changes
3. In the test I capture a new screenshot
4. I compare the dimensions of  the new screenshot to the previous screenshot. If they don't match I fail the test
5. I compare each pixel of the expected image to the new image. If the pixel colors match, I write that pixel to a new image (diff image). If the pixels don't match, I write a red pixel to the diff image
6. After comparing all pixels, I get the percentage of red pixels in the diff image. If the percentage is higher than a user defined threshold, I attach the expected image, the actual image, and the diff image to the report and fail the test.

If there are expected visual changes, I just run the test and replace the expected image with the new image. It's actual very reliable and doesn't take much maintenance.

**Reply by Nerwen18** [reply]
Hello :) thank you for your suggestion, I'll give it a try!
